apps/astarte_appengine_api/lib/astarte_appengine_api/device/queries.ex

```diff
@@ -365,68 +349,63 @@ defmodule Astarte.AppEngine.API.Device.Queries do
-        if column_atoms[obj_key] != nil do
-          column_name = CQLUtils.endpoint_to_db_column_name(obj_key)
-
-          db_value = to_db_friendly_type(obj_value)
-          next_query_values_acc = Map.put(query_values_acc, column_atoms[obj_key], db_value)
-          next_placeholders_acc = "#{placeholders_acc} :#{to_string(column_atoms[obj_key])},"
-          next_query_acc = "#{query_acc} #{column_name}, "
-
-          {next_query_values_acc, next_placeholders_acc, next_query_acc}
-        else
-          Logger.warning(
-            "Unexpected object key #{inspect(obj_key)} with value #{inspect(obj_value)}."
-          )
+    {timestamp_ms, timestamp_submillis} = split_ms_and_submillis(timestamp)
 
-          query_values_acc
-        end
-      end)
+    base_attributes = %{
+      device_id: device_id,
+      path: path
+    }
 
-    {query_columns, placeholders} =
-      if explicit_timestamp do
-        {"value_timestamp, #{query_columns}", ":value_timestamp, #{placeholders}"}
+    timestamp_attributes =
+      if explicit_timestamp? do
+        %{
+          value_timestamp: timestamp_ms,
+          reception_timestamp: timestamp_ms,
+          reception_timestamp_submillis: timestamp_submillis
+        }
       else
-        {query_columns, placeholders}
+        %{reception_timestamp: timestamp_ms, reception_timestamp_submillis: timestamp_submillis}
       end
 
-    {timestamp_ms, timestamp_submillis} = split_ms_and_submillis(timestamp)
+    value =
+      value
+      |> Enum.flat_map(fn {key, value} ->
+        # filter map
+        case Map.fetch(column_meta, key) do
+          {:ok, meta} ->
+            %{name: name, type: type} = meta
+            data = %{type: type, value: value}
+            [{name, data}]
 
-    # TODO: :reception_timestamp_submillis is just a place holder right now
-    insert_query =
-      DatabaseQuery.new()
-      |> DatabaseQuery.statement("""
-      INSERT INTO #{interface_descriptor.storage} (device_id, path, #{query_columns} reception_timestamp, reception_timestamp_submillis)
-        VALUES (:device_id, :path, #{placeholders} :reception_timestamp, :reception_timestamp_submillis) #{ttl_string};
-      """)
-      |> DatabaseQuery.put(:device_id, device_id)
-      |> DatabaseQuery.put(:path, path)
-      |> DatabaseQuery.put(:value_timestamp, timestamp_ms)
-      |> DatabaseQuery.put(:reception_timestamp, timestamp_ms)
-      |> DatabaseQuery.put(:reception_timestamp_submillis, timestamp_submillis)
-      |> DatabaseQuery.merge(query_values)
+          :error ->
+            Logger.warning("Unexpected object key #{inspect(key)} with value #{inspect(value)}.")
 
-    # TODO: |> DatabaseQuery.consistency(insert_consistency(interface_descriptor, endpoint))
+            []
+        end
+      end)
 
-    DatabaseQuery.call!(db_client, insert_query)
+    value_attributes = value |> Map.new(fn {column, data} -> {column, data.value} end)
```

Comment:
Nit: what about moving this to a new function, in order to separate the value preparation part and decrease the cognitive load of reading this chunk of code?